File: Leetcode_POTD_2025/7_JULY/3440_Reschedule_Meetings_for_Maximum_Free_Time_II.py
class Solution(object):
    def maxFreeTime(self, eventTime, startTime, endTime):
        """
        :type eventTime: int
        :type startTime: List[int]
        :type endTime: List[int]
        :rtype: int
        """
        n = len(startTime)
        p, q, mx = 0, startTime[0], 0
        h = [startTime[i] - endTime[i - 1] for i in range(1, n)]
        h.insert(0, q)
        h.append(eventTime - endTime[-1])

        suf = [0] * len(h)
        for i in range(len(h) - 2, -1, -1):
            suf[i] = max(suf[i + 1], h[i + 1])

        for i in range(1, len(h)):
            dur = endTime[i - 1] - startTime[i - 1]
            if dur <= max(p, suf[i]):
                mx = max(mx, h[i - 1] + h[i] + dur)
            mx = max(mx, h[i - 1] + h[i])
            p = max(p, h[i - 1])
        return mx

File: Leetcode_POTD_2025/7_JULY/3440_Reschedule_Meetings_for_Maximum_Free_Time_II.cpp
class Solution {
public:
    int maxFreeTime(int z, vector<int>& a, vector<int>& b) {
        int n = a.size(), p = 0, q = a[0], mx = 0;
        vector<int> h;
        for (int i = 1; i < n; ++i) h.push_back(a[i] - b[i - 1]);
        h.insert(h.begin(), q), h.push_back(z - b[n - 1]);

        vector<int> suf(h.size()); suf.back() = 0;
        for (int i = h.size() - 2; i >= 0; --i) suf[i] = max(suf[i + 1], h[i + 1]);

        for (int i = 1; i < h.size(); ++i) {
            int dur = b[i - 1] - a[i - 1];
            if (dur <= max(p, suf[i])) mx = max(mx, h[i - 1] + h[i] + dur);
            mx = max(mx, h[i - 1] + h[i]);
            p = max(p, h[i - 1]);
        }
        return mx;
    }
};
